migrator/zenodo_rdm_migrator/actions/transform/files.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# Invenio-RDM-Migrator is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Invenio RDM migration files actions module."""
+
+
+from invenio_rdm_migrator.actions import TransformAction
+from invenio_rdm_migrator.load.postgresql.transactions.operations import OperationType
+from invenio_rdm_migrator.streams.actions import load
+
+
+class FileUploadAction(TransformAction):
+    """Zenodo to RDM file upload action."""
+
+    name = "upload-file"
+    load_cls = load.FileUploadAction
+
+    @classmethod
+    def matches_action(cls, tx):
+        """Checks if the data corresponds with that required by the action.
+
+        Should contain two updates on the bucket (updated, size), the creation
+        and the link to the file_id (update) on the object version, and the creation and commit
+        (update) of the file instance.
+        """
+        rules = {
+            "files_bucket": [OperationType.UPDATE, OperationType.UPDATE],
+            "files_object": [OperationType.UPDATE, OperationType.INSERT],  # lifo
+            "files_files": [OperationType.UPDATE, OperationType.INSERT],  # lifo
+        }
+
+        for operation in tx.operations:
+            table_name = operation["source"]["table"]
+            try:
+                rule = rules[table_name].pop()
+                if not rule == operation["op"]:
+                    return False
+            except (KeyError, IndexError):
+                return False
+
+        for ops in rules.values():  # check that rules were exhausted
+            if len(ops) > 0:
+                return False
+
+        return True
+
+    def _transform_data(self):
+        """Transforms the data and returns an instance of the mapped_cls."""
+        bucket = {}
+        ov = {}
+        fi = {}
+        for operation in self.tx.operations:
+            table_name = operation["source"]["table"]
+            self._microseconds_to_isodate(
+                data=operation["after"], fields=["created", "updated"]
+            )
+
+            if table_name == "files_bucket":
+                bucket = {**bucket, **operation["after"]}
+            elif table_name == "files_object":
+                ov = {**ov, **operation["after"]}
+            elif table_name == "files_files":
+                fi = {**fi, **operation["after"]}
+
+        fr = {
+            "id": None,  # generated by the load action
+            "json": {},
+            "created": ov["created"],
+            "updated": ov["updated"],
+            "version_id": 1,
+            "key": ov["key"],
+            "record_id": None,  # calculated by the load action
```

Comment:
minor/comment: do we have a rule of thumb over who is responsible for the "linking" to existing objects? IIRC, we do have access to the state in the Transform In this case, so we could fish the record ID using the bucket ID.

Reply:
I was also divided at heart on this one. I decided to keep it in the load, since it is how we do for other cases in drafts. 

For example, both in snapshot and action migration we might have to check if it is a draft of a published version or not; then we have to use one data or another. At the moment this is taken from the state and is done in load.

In my opinion, it is the same type of operation: _calculate a value based on the state_. It falls again into is it business logic/data transformation of is it data processing? To be consistent I kept it in the `load`, but I'm not against moving all this operations to transform. I think it makes sense:
- load: get data and db interaction
- tranform: do all transformation (from data and state)

⚠️ this would be a big refactor 😢

Reply:
Thanks for the mental walkthrough :)

I retraced a bit what we were discussing at the very beginning of the ETL pipeline and I think it's good practice to delegate to the Load. 

The logic is that you can feed to the Load step data that has "holes" (e.g. missing PKs, relations/FKs) and it will take the responsibility of either "minting" new PKs or figuring out if/what is the related object FK. With that in mind, in this example the data we're passing to Load is saying "there was a file uploaded to bucket X, that is connected to a record; please figure out which record that is".

What I would maybe make a bit more clear throughout the classes is that this was a `DraftFileUpload` action (and not just any `FileUpload` action). I think with the type of actions it's generally better to be very specific (since they have to represent real-life possible actions on either legacy or RDM). This doesn't mean of course that the underlying code cannot be generic and shared (i.e. no need to duplicate code for the Transform/Load actions).